hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -622,7 +623,31 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
     val rightKeyFieldIdx = rightRVDType.kFieldIdx
     val leftValueFieldIdx = leftRVDType.valueFieldIdx
     val rightValueFieldIdx = rightRVDType.valueFieldIdx
-    val newRowPType = PType.canonical(newRowType).asInstanceOf[PStruct]
+
+    val leftKeyType = leftRVDType.kType
+    var leftValueType = leftRVDType.valueType
+    var rightValueType = rightRVDType.valueType
```

Comment:
var + mutation is a less clear pattern than having the match return a triple of (key fields, left fields, right fields). These can all have type `Array[(String, PType)]`

Reply:
also, we don't need the setFieldsRequiredness method on PCanonicalStruct, it  adds complexity without good reason

Reply:
> var + mutation is a less clear pattern than having the match return a triple of (key fields, left fields, right fields). These can all have type Array[(String, PType)]

I disagree. Tried both ways this was more ergonomic.

Reply:
you may have saved a couple lines of code, but it also means the reader needs to look in multiple places to be convinced that a rule for a specific join type is correct

Reply:
> you may have saved a couple lines of code, but it also means the reader needs to look 

It's a standard pattern: initialize variables outside of the block where their values are set. 

> also, we don't need the setFieldsRequiredness method on PCanonicalStruct

Without having this on PCanonicalStruct we need to cast to PCanonicalStruct, because we don't have a "copy" constructor on ptypes.

Reply:
> Without having this on PCanonicalStruct we need to cast to PCanonicalStruct

We don't actually need a struct at all! We should get rid of `++` on PStruct entirely, this code should construct a `PCanonicalStruct` passing in fields.

Reply:
Ah true can do this :)

Reply:
Basically all the abstract methods on PStruct that modify the type need to go. update/appendKey, ++, typeAfterSelect, etc.

Reply:
I thought we wouldn’t want to modify the rvd row type, just subset it, but if we don’t mind that the strict will become canonical in the future when we have non-canonical structs, then your suggestion works.

Reply:
addressed assuming your opinion doesn't change